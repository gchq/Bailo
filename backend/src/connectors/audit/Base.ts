import { Request } from 'express'

import { AccessRequestDoc } from '../../models/AccessRequest.js'
import { FileInterface, FileInterfaceDoc } from '../../models/File.js'
import { InferenceDoc } from '../../models/Inference.js'
import { ModelCardInterface, ModelDoc, ModelInterface } from '../../models/Model.js'
import { ReleaseDoc } from '../../models/Release.js'
import { ResponseInterface } from '../../models/Response.js'
import { ReviewInterface } from '../../models/Review.js'
import { SchemaDoc, SchemaInterface } from '../../models/Schema.js'
import { TokenDoc } from '../../models/Token.js'
import { ModelSearchResult } from '../../routes/v2/model/getModelsSearch.js'
import { BailoError } from '../../types/error.js'

const AuditKind = {
  Create: 'Create',
  View: 'View',
  Update: 'Update',
  Delete: 'Delete',
  Search: 'Search',
} as const
export type AuditKindKeys = (typeof AuditKind)[keyof typeof AuditKind]

export const AuditInfo = {
  CreateModel: { typeId: 'CreateModel', description: 'Model Created', auditKind: AuditKind.Create },
  ViewModel: { typeId: 'ViewModel', description: 'Model Viewed', auditKind: AuditKind.View },
  UpdateModel: { typeId: 'UpdateModel', description: 'Model Updated', auditKind: AuditKind.Update },
  SearchModels: { typeId: 'SearchModels', description: 'Model Searched', auditKind: AuditKind.Search },

  CreateModelCard: { typeId: 'CreateModelCard', description: 'Model Card Created', auditKind: AuditKind.Create },
  ViewModelCard: { typeId: 'ViewModelCard', description: 'Model Card Viewed', auditKind: AuditKind.View },
  ViewModelCardRevisions: {
    typeId: 'ViewModelCardRevisions',
    description: 'Model Card Revisions Viewed',
    auditKind: AuditKind.View,
  },
  UpdateModelCard: { typeId: 'UpdateModelCard', description: 'Model Card Updated', auditKind: AuditKind.Update },

  CreateFile: { typeId: 'CreateFile', description: 'File Information Created', auditKind: AuditKind.Create },
  ViewFile: { typeId: 'ViewFile', description: 'File Downloaded', auditKind: AuditKind.View },
  ViewFiles: { typeId: 'ViewFiles', description: 'File Information Viewed', auditKind: AuditKind.View },
  DeleteFile: { typeId: 'DeleteFile', description: 'File Information Deleted', auditKind: AuditKind.Delete },

  CreateRelease: { typeId: 'CreateRelease', description: 'Release Created', auditKind: AuditKind.Create },
  ViewRelease: { typeId: 'ViewRelease', description: 'Release Viewed', auditKind: AuditKind.View },
  UpdateRelease: { typeId: 'UpdateRelease', description: 'Release Updated', auditKind: AuditKind.Update },
  DeleteRelease: { typeId: 'DeleteRelease', description: 'Release Deleted', auditKind: AuditKind.Delete },
  ViewReleases: { typeId: 'ViewReleases', description: 'Releases Viewed', auditKind: AuditKind.View },

  CreateUserToken: { typeId: 'CreateUserToken', description: 'Token Created', auditKind: AuditKind.Create },
  ViewUserTokens: { typeId: 'ViewUserToken', description: 'Token Viewed', auditKind: AuditKind.View },
  DeleteUserToken: { typeId: 'DeleteUserToken', description: 'Token Deleted', auditKind: AuditKind.Delete },

  CreateAccessRequest: {
    typeId: 'CreateAccessRequest',
    description: 'Access Request Created',
    auditKind: AuditKind.Create,
  },
  ViewAccessRequest: { typeId: 'ViewAccessRequest', description: 'Access Request Viewed', auditKind: AuditKind.View },
  UpdateAccessRequest: {
    typeId: 'UpdateAccess Request',
    description: 'Access Request Updated',
    auditKind: AuditKind.Update,
  },
  DeleteAccessRequest: {
    typeId: 'UpdateAccessRequest',
    description: 'Access Request Deleted',
    auditKind: AuditKind.Update,
  },
  ViewAccessRequests: {
    typeId: 'ViewAccessRequests',
    description: 'Access Requests Viewed',
    auditKind: AuditKind.View,
  },

  SearchReviews: {
    typeId: 'SearchReviews',
    description: 'Reviews Searched',
    auditKind: AuditKind.Search,
  },
  CreateReviewResponse: {
    typeId: 'CreateReviewResponse',
    description: 'Review Response Created',
    auditKind: AuditKind.Create,
  },
  UpdateReviewResponse: {
    typeId: 'UpdateReviewResponse',
    description: 'Review Response Updated',
    auditKind: AuditKind.Update,
  },

  CreateSchema: { typeId: 'CreateSchema', description: 'Schema Created', auditKind: AuditKind.Create },
  SearchSchemas: { typeId: 'SearchedSchemas', description: 'Schemas Searched', auditKind: AuditKind.Search },
  ViewSchema: { typeId: 'ViewSchema', description: 'Schema Viewed', auditKind: AuditKind.View },
  DeleteSchema: { typeId: 'DeleteSchema', description: 'Schema Deleted', auditKind: AuditKind.Delete },
  UpdateSchema: { typeId: 'UpdateSchema', description: 'Schema Updated', auditKind: AuditKind.Update },

  ViewModelImages: { typeId: 'ViewModelImages', description: 'Model Images Viewed', auditKind: AuditKind.View },

  CreateInference: { typeId: 'CreateInference', description: 'Inference Service Created', auditKind: AuditKind.Create },
  UpdateInference: { typeId: 'UpdateInference', description: 'Inference Service Updated', auditKind: AuditKind.Update },
  ViewInference: { typeId: 'ViewInference', description: 'Inference Service Viewed', auditKind: AuditKind.View },
  ViewInferences: { typeId: 'ViewInferences', description: 'Inferences Viewed', auditKind: AuditKind.View },

  CreateExport: { typeId: 'CreateExport', description: 'Model Exported', auditKind: AuditKind.Create },
  CreateImport: { typeId: 'CreateImport', description: 'Model Imported', auditKind: AuditKind.Create },

  ViewResponses: {
    typeId: 'ViewResponses',
    description: 'View a list of responses',
    auditKind: AuditKind.View,
  },
  CreateResponse: {
    typeId: 'CreateResponse',
    description: 'Review or comment responses created',
    auditKind: AuditKind.Create,
  },
  UpdateResponse: {
    typeId: 'UpdateResponse',
    description: 'Updated a comment or review response',
    auditKind: AuditKind.Update,
  },
} as const
export type AuditInfoKeys = (typeof AuditInfo)[keyof typeof AuditInfo]

export abstract class BaseAuditConnector {
  abstract onCreateModel(req: Request, model: ModelDoc)
  abstract onViewModel(req: Request, model: ModelDoc)
  abstract onUpdateModel(req: Request, model: ModelDoc)
  abstract onSearchModel(req: Request, models: ModelSearchResult[])

  abstract onCreateModelCard(req: Request, modelId: string, modelCard: ModelCardInterface)
  abstract onViewModelCard(req: Request, modelId: string, modelCard: ModelCardInterface)
  abstract onUpdateModelCard(req: Request, modelId: string, modelCard: ModelCardInterface)
  abstract onViewModelCardRevisions(req: Request, modelId: string, modelCards: ModelCardInterface[])

  abstract onCreateFile(req: Request, file: FileInterfaceDoc)
  abstract onViewFile(req: Request, file: FileInterfaceDoc)
  abstract onViewFiles(req: Request, modelId: string, files: FileInterface[])
  abstract onDeleteFile(req: Request, modelId: string, fileId: string)

  abstract onCreateRelease(req: Request, release: ReleaseDoc)
  abstract onViewRelease(req: Request, release: ReleaseDoc)
  abstract onUpdateRelease(req: Request, release: ReleaseDoc)
  abstract onDeleteRelease(req: Request, modelId: string, semver: string)
  abstract onViewReleases(req: Request, releases: ReleaseDoc[])

  abstract onCreateCommentResponse(req: Request, response: ResponseInterface)
  abstract onCreateReviewResponse(req: Request, response: ResponseInterface)

  abstract onViewResponses(req: Request, responseInterfaces: ResponseInterface[])
  abstract onUpdateResponse(req: Request, responseId: string)

  abstract onCreateUserToken(req: Request, token: TokenDoc)
  abstract onViewUserTokens(req: Request, tokens: TokenDoc[])
  abstract onDeleteUserToken(req: Request, accessKey: string)

  abstract onCreateAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onViewAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onUpdateAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onDeleteAccessRequest(req: Request, accessRequestId: string)
  abstract onViewAccessRequests(req: Request, accessRequests: AccessRequestDoc[])

  abstract onSearchReviews(req: Request, reviews: (ReviewInterface & { model: ModelInterface })[])

  abstract onSearchSchemas(req: Request, schemas: SchemaInterface[])
  abstract onCreateSchema(req: Request, schema: SchemaInterface)
  abstract onViewSchema(req: Request, schema: SchemaInterface)
  abstract onDeleteSchema(req: Request, schemaId: string)
  abstract onUpdateSchema(req: Request, schema: SchemaDoc)

  abstract onCreateInference(req: Request, inference: InferenceDoc)
  abstract onUpdateInference(req: Request, inference: InferenceDoc)
  abstract onViewInference(req: Request, inference: InferenceDoc)
  abstract onViewInferences(req: Request, inference: InferenceDoc[])

  abstract onViewModelImages(
    req: Request,
    modelId: string,
    images: { repository: string; name: string; tags: string[] }[],
  )

  abstract onCreateS3Export(req: Request, modelId: string, semvers?: string[])
  abstract onCreateImport(
    req: Request,
    mirroredModelId: string,
    sourceModelId: string,
    modelCardVersions: number[],
    exporter: string,
  )

  abstract onError(req: Request, error: BailoError)

  checkEventType(auditInfo: AuditInfoKeys, req: Request) {
    if (auditInfo.typeId !== req.audit.typeId && auditInfo.description !== req.audit.description) {
      throw new Error(`Audit: Expected type '${JSON.stringify(auditInfo)}' but recieved '${JSON.stringify(req.audit)}'`)
    }
  }
}
