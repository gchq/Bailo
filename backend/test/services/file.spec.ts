import { Readable } from 'stream'
import { describe, expect, test, vi } from 'vitest'

import { FileAction } from '../../src/connectors/authorisation/actions.js'
import authorisation from '../../src/connectors/authorisation/index.js'
import { FileScanResult, ScanState } from '../../src/connectors/fileScanning/Base.js'
import { UserInterface } from '../../src/models/User.js'
import {
  downloadFile,
  getFilesByIds,
  getFilesByModel,
  getTotalFileSize,
  isFileInterfaceDoc,
  removeFile,
  rerunFileScan,
  uploadFile,
} from '../../src/services/file.js'

vi.mock('../../src/connectors/authorisation/index.js')
vi.mock('../../src/connectors/fileScanning/index.js')

const logMock = vi.hoisted(() => ({
  info: vi.fn(),
  warn: vi.fn(),
}))
vi.mock('../../src/services/log.js', async () => ({
  default: logMock,
}))

const configMock = vi.hoisted(
  () =>
    ({
      avScanning: {
        clamdscan: {
          host: 'test',
          port: 8080,
        },
      },
      s3: {
        buckets: {
          uploads: 'uploads',
          registry: 'registry',
        },
      },
      connectors: {
        fileScanners: {
          kinds: ['clamAV'],
          retryDelayInMinutes: 5,
          maxInitRetries: 5,
          initRetryDelay: 5000,
        },
      },
    }) as any,
)
vi.mock('../../src/utils/config.js', () => ({
  __esModule: true,
  default: configMock,
}))

const fileScanResult: FileScanResult = {
  state: 'complete',
  isInfected: false,
  toolName: 'Test',
  lastRunAt: new Date(),
}

const fileScanningMock = vi.hoisted(() => ({
  info: vi.fn(() => []),
  scan: vi.fn(() => new Promise(() => [fileScanResult])),
}))
vi.mock('../../src/connectors/fileScanning/index.js', async () => ({ default: fileScanningMock }))

const s3Mocks = vi.hoisted(() => ({
  putObjectStream: vi.fn(() => ({ fileSize: 100 })),
  getObjectStream: vi.fn(() => ({ Body: { pipe: vi.fn() } })),
}))
vi.mock('../../src/clients/s3.js', () => s3Mocks)

const modelMocks = vi.hoisted(() => ({
  getModelById: vi.fn(() => ({ settings: { mirror: { sourceModelId: '' } } })),
}))
vi.mock('../../src/services/model.js', () => modelMocks)

const releaseServiceMocks = vi.hoisted(() => ({
  removeFileFromReleases: vi.fn(),
}))
vi.mock('../../src/services/release.js', () => releaseServiceMocks)

const testFileId = '73859F8D26679D2E52597326'

const fileModelMocks = vi.hoisted(() => {
  const obj: any = {}

  obj.aggregate = vi.fn(() => obj)
  obj.match = vi.fn(() => obj)
  obj.group = vi.fn(() => obj)

  obj.save = vi.fn(() => obj)
  obj.find = vi.fn(() => obj)
  obj.delete = vi.fn(() => obj)
  obj.findOneAndDelete = vi.fn(() => obj)

  obj.toObject = vi.fn(() => obj)

  obj._id = vi.fn()

  const model: any = vi.fn(() => obj)
  Object.assign(model, obj)

  return model
})
vi.mock('../../src/models/File.js', () => ({ default: fileModelMocks }))

const scanModelMocks = vi.hoisted(() => {
  const obj: any = {}

  obj.aggregate = vi.fn(() => obj)

  obj.find = vi.fn(() => obj)
  obj.findOne = vi.fn(() => obj)
  obj.delete = vi.fn(() => obj)

  const model: any = vi.fn(() => obj)
  Object.assign(model, obj)

  return model
})
vi.mock('../../src/models/Scan.js', () => ({ default: scanModelMocks }))

const baseScannerMock = vi.hoisted(() => ({
  ScanState: {
    NotScanned: 'notScanned',
    InProgress: 'inProgress',
    Complete: 'complete',
    Error: 'error',
  },
}))
vi.mock('../../src/connectors/filescanning/Base.js', () => baseScannerMock)

const clamscan = vi.hoisted(() => ({ on: vi.fn() }))
vi.mock('clamscan', () => ({
  __esModule: true,
  default: vi.fn(() => ({ init: vi.fn(() => ({ passthrough: vi.fn(() => clamscan) })) })),
}))

describe('services > file', () => {
  test('uploadFile > success', async () => {
    const user = { dn: 'testUser' } as UserInterface
    const modelId = 'testModelId'
    const name = 'testFile'
    const mime = 'text/plain'
    const stream = new Readable() as any
    const result = await uploadFile(user, modelId, name, mime, stream)

    expect(s3Mocks.putObjectStream).toBeCalled()
    expect(fileModelMocks.save).toBeCalled()
    expect(result).toMatchSnapshot()
  })

  test('uploadFile > virus scan initialised', async () => {
    vi.spyOn(configMock, 'avScanning', 'get').mockReturnValue({ clamdscan: 'test' })
    vi.spyOn(configMock, 'connectors', 'get').mockReturnValue({
      fileScanners: {
        kinds: ['clamAV'],
      },
    })
    const user = { dn: 'testUser' } as UserInterface
    const modelId = 'testModelId'
    const name = 'testFile'
    const mime = 'text/plain'
    const stream = new Readable() as any

    const result = await uploadFile(user, modelId, name, mime, stream)

    expect(s3Mocks.putObjectStream).toBeCalled()
    expect(fileModelMocks.save).toBeCalled()
    expect(result).toMatchSnapshot()
    expect(clamscan.on.mock.calls).toMatchSnapshot()
  })

  test('uploadFile > no permission', async () => {
    vi.mocked(authorisation.file).mockResolvedValue({
      info: 'You do not have permission to upload a file to this model.',
      success: false,
      id: '',
    })

    expect(() => uploadFile({} as any, 'modelId', 'name', 'mime', new Readable() as any)).rejects.toThrowError(
      /^You do not have permission to upload a file to this model./,
    )
  })

  test('uploadFile > should throw an error when attempting to upload a file to a mirrored model', async () => {
    const user = { dn: 'testUser' } as UserInterface
    const modelId = 'testModelId'
    const name = 'testFile'
    const mime = 'text/plain'
    const stream = new Readable() as any

    vi.mocked(authorisation.file).mockResolvedValue({
      info: 'Cannot upload files to a mirrored model.',
      success: false,
      id: '',
    })

    expect(() => uploadFile(user, modelId, name, mime, stream)).rejects.toThrowError(
      /^Cannot upload files to a mirrored model./,
    )
    expect(fileModelMocks.save).not.toBeCalled()
  })
  test('removeFile > success', async () => {
    const user = { dn: 'testUser' } as UserInterface
    const modelId = 'testModelId'

    fileModelMocks.aggregate.mockResolvedValueOnce([
      { modelId: 'testModel', _id: { toString: vi.fn(() => testFileId) } },
    ])

    const result = await removeFile(user, modelId, testFileId)

    expect(releaseServiceMocks.removeFileFromReleases).toBeCalled()
    expect(result).toMatchSnapshot()
  })

  test('removeFile > no release permission', async () => {
    const user = { dn: 'testUser' } as UserInterface
    const modelId = 'testModelId'

    fileModelMocks.aggregate.mockResolvedValueOnce([
      { modelId: 'testModel', _id: { toString: vi.fn(() => testFileId) } },
    ])

    releaseServiceMocks.removeFileFromReleases.mockRejectedValueOnce('Cannot update releases')

    const result = removeFile(user, modelId, testFileId)

    expect(result).rejects.toThrowError(/^Cannot update releases/)
    expect(fileModelMocks.delete).not.toBeCalled()
  })

  test('removeFile > no file permission', async () => {
    fileModelMocks.aggregate.mockResolvedValueOnce([
      { modelId: 'testModel', _id: { toString: vi.fn(() => testFileId) } },
    ])
    vi.mocked(authorisation.file).mockImplementation(async (_user, _model, _file, action) => {
      if (action === FileAction.View) return { success: true, id: '' }
      if (action === FileAction.Delete)
        return { success: false, info: 'You do not have permission to delete a file from this model.', id: '' }

      return { success: false, info: 'Unknown action.', id: '' }
    })

    const user = { dn: 'testUser' } as UserInterface
    const modelId = 'testModelId'

    expect(() => removeFile(user, modelId, testFileId)).rejects.toThrowError(
      /^You do not have permission to delete a file from this model./,
    )
    expect(fileModelMocks.delete).not.toBeCalled()
  })

  test('removeFile > should throw an error when attempting to remove a file from a mirrored model', async () => {
    const user = { dn: 'testUser' } as UserInterface
    const modelId = 'testModelId'

    fileModelMocks.aggregate.mockResolvedValueOnce([
      { modelId: 'testModel', _id: { toString: vi.fn(() => testFileId) } },
    ])
    vi.mocked(authorisation.file).mockResolvedValue({
      info: 'Cannot remove file from a mirrored model.',
      success: false,
      id: '',
    })
    expect(() => removeFile(user, modelId, testFileId)).rejects.toThrowError(
      /^Cannot remove file from a mirrored model./,
    )
    expect(fileModelMocks.delete).not.toBeCalled()
  })

  test('getFilesByModel > success', async () => {
    fileModelMocks.aggregate.mockResolvedValueOnce([{ example: 'file', _id: { toString: vi.fn(() => testFileId) } }])

    const user = { dn: 'testUser' } as UserInterface
    const modelId = 'testModelId'

    const result = await getFilesByModel(user, modelId)

    expect(result).toMatchSnapshot()
  })

  test('getFilesByModel > no permission', async () => {
    vi.mocked(authorisation.files).mockResolvedValue([
      {
        info: 'You do not have permission to get the files from this model.',
        success: false,
        id: '',
      },
    ])

    fileModelMocks.aggregate.mockResolvedValueOnce([{ example: 'file', _id: { toString: vi.fn(() => testFileId) } }])

    const user = { dn: 'testUser' } as UserInterface
    const modelId = 'testModelId'

    const files = await getFilesByModel(user, modelId)
    expect(files).toStrictEqual([])
  })

  test('getFilesByIds > success', async () => {
    fileModelMocks.aggregate.mockResolvedValueOnce([
      { example: 'file', avScan: [], _id: { toString: vi.fn(() => testFileId) } },
    ])

    const user = { dn: 'testUser' } as UserInterface
    const modelId = 'testModelId'
    const fileIds = [testFileId]

    const files = await getFilesByIds(user, modelId, fileIds)

    expect(files).toMatchSnapshot()
  })

  test('getFilesByIds > success with scans mapped', async () => {
    fileModelMocks.aggregate.mockResolvedValueOnce([
      {
        example: 'file',
        _id: '123',
        avScan: [{ fileId: '123' }, { fileId: '123' }],
      },
      {
        example: 'file',
        _id: '321',
        avScan: [{ fileId: '321' }],
      },
    ])
    vi.mocked(authorisation.files).mockResolvedValue([
      { success: true, id: '123' },
      { success: true, id: '321' },
    ])

    const user = { dn: 'testUser' } as UserInterface
    const modelId = 'testModelId'
    const fileIds = ['123', '321']

    const files = await getFilesByIds(user, modelId, fileIds)

    expect(files).toMatchSnapshot()
  })

  test('getFilesByIds > no file ids', async () => {
    fileModelMocks.aggregate.mockResolvedValueOnce([{ example: 'file', _id: { toString: vi.fn(() => testFileId) } }])

    const user = { dn: 'testUser' } as UserInterface
    const modelId = 'testModelId'
    const fileIds = []

    const files = await getFilesByIds(user, modelId, fileIds)

    expect(files).toStrictEqual(fileIds)
  })

  test('getFilesByIds > files not found', async () => {
    fileModelMocks.aggregate.mockResolvedValueOnce([{ example: 'file', _id: { toString: vi.fn(() => testFileId) } }])

    const user = { dn: 'testUser' } as UserInterface
    const modelId = 'testModelId'
    const fileIds = [testFileId, testFileId]

    const files = getFilesByIds(user, modelId, fileIds)

    expect(files).rejects.toThrowError(/^The requested files were not found./)
  })

  test('getFilesByIds > no permission', async () => {
    vi.mocked(authorisation.files).mockResolvedValue([
      {
        info: 'You do not have permission to get the files from this model.',
        success: false,
        id: '',
      },
    ])
    fileModelMocks.aggregate.mockResolvedValueOnce([
      { example: 'file', avScan: [], _id: { toString: vi.fn(() => testFileId) } },
    ])

    const user = { dn: 'testUser' } as UserInterface
    const modelId = 'testModelId'
    const fileIds = [testFileId]

    const files = await getFilesByIds(user, modelId, fileIds)
    expect(files).toStrictEqual([])
  })

  test('downloadFile > success', async () => {
    const user = { dn: 'testUser' } as UserInterface
    const range = { start: 0, end: 50 }

    fileModelMocks.aggregate.mockResolvedValueOnce([
      { modelId: 'testModel', _id: { toString: vi.fn(() => testFileId) } },
    ])

    const result = await downloadFile(user, testFileId, range)

    expect(result).toMatchSnapshot()
  })

  test('downloadFile > no permission', async () => {
    const user = { dn: 'testUser' } as UserInterface
    const range = { start: 0, end: 50 }
    fileModelMocks.aggregate.mockResolvedValueOnce([
      { modelId: 'testModel', _id: { toString: vi.fn(() => testFileId) } },
    ])

    vi.mocked(authorisation.file).mockImplementation(async (_user, _model, _file, action) => {
      if (action === FileAction.View) return { success: true, id: '' }
      if (action === FileAction.Download)
        return { success: false, info: 'You do not have permission to download this model.', id: '' }

      return { success: false, info: 'Unknown action.', id: '' }
    })

    await expect(downloadFile(user, testFileId, range)).rejects.toThrowError(
      /^You do not have permission to download this model./,
    )
  })

  test('getTotalFileSize > returns file size', async () => {
    fileModelMocks.group.mockResolvedValueOnce([{ totalSize: 42 }])
    const size = await getTotalFileSize(['1', '2', '3'])

    expect(size).toBe(42)
  })

  test('rerunFileScan > successfully reruns a file scan', async () => {
    const createdAtTimeInMilliseconds = new Date().getTime() - 2000000
    fileModelMocks.aggregate.mockResolvedValueOnce([
      {
        name: 'file.txt',
        size: 123,
        _id: { toString: vi.fn(() => testFileId) },
      },
    ])
    scanModelMocks.find.mockResolvedValueOnce([
      {
        state: ScanState.Complete,
        lastRunAt: new Date(createdAtTimeInMilliseconds),
      },
    ])
    const scanStatus = await rerunFileScan({} as any, 'model123', testFileId)
    expect(scanStatus).toBe('Scan started for file.txt')
  })

  test('rerunFileScan > throws bad request when attempting to upload an empty file', async () => {
    fileModelMocks.aggregate.mockResolvedValueOnce([
      {
        name: 'file.txt',
        size: 0,
        _id: { toString: vi.fn(() => testFileId) },
      },
    ])
    scanModelMocks.find.mockResolvedValueOnce([
      {
        state: ScanState.Complete,
      },
    ])
    await expect(rerunFileScan({} as any, 'model123', testFileId)).rejects.toThrowError(
      /^Cannot run scan on an empty file/,
    )
  })

  test('rerunFileScan > does not rerun file scan before delay is over', async () => {
    fileModelMocks.aggregate.mockResolvedValueOnce([
      {
        name: 'file.txt',
        size: 123,
        _id: { toString: vi.fn(() => testFileId) },
      },
    ])
    scanModelMocks.find.mockResolvedValueOnce([{ state: ScanState.Complete, lastRunAt: new Date() }])
    await expect(rerunFileScan({} as any, 'model123', testFileId)).rejects.toThrowError(
      /^Please wait 5 minutes before attempting a rescan file.txt/,
    )
  })

  test('isFileInterfaceDoc > success', async () => {
    const result = isFileInterfaceDoc({
      modelId: '',
      name: '',
      size: 1,
      mime: '',
      bucket: '',
      path: '',
      complete: true,
      deleted: false,
      createdAt: '',
      updatedAt: '',
      _id: '',
    })

    expect(result).toBe(true)
  })

  test('isFileInterfaceDoc > missing property', async () => {
    const result = isFileInterfaceDoc({
      modelId: '',
      name: '',
      size: 1,
      mime: '',
      path: '',
      complete: true,
      deleted: false,
      createdAt: '',
      updatedAt: '',
      _id: '',
    })

    expect(result).toBe(false)
  })

  test('isFileInterfaceDoc > wrong type', async () => {
    const result = isFileInterfaceDoc(null)

    expect(result).toBe(false)
  })
})
