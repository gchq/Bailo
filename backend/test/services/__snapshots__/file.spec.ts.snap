// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`services > file > downloadFile > success 1`] = `
{
  "Body": {
    "pipe": [MockFunction spy],
  },
}
`;

exports[`services > file > getFilesByIds > success 1`] = `
[
  {
    "_id": {
      "toString": [MockFunction spy],
    },
    "avScan": [],
    "example": "file",
  },
]
`;

exports[`services > file > getFilesByIds > success with scans mapped 1`] = `
[
  {
    "_id": "123",
    "avScan": [
      {
        "fileId": "123",
      },
      {
        "fileId": "123",
      },
    ],
    "example": "file",
  },
  {
    "_id": "321",
    "avScan": [
      {
        "fileId": "321",
      },
    ],
    "example": "file",
  },
]
`;

exports[`services > file > getFilesByModel > success 1`] = `
[
  {
    "_id": {
      "toString": [MockFunction spy],
    },
    "example": "file",
  },
]
`;

exports[`services > file > removeFile > success 1`] = `
{
  "_id": {
    "toString": [MockFunction spy] {
      "calls": [
        [],
      ],
      "results": [
        {
          "type": "return",
          "value": "73859F8D26679D2E52597326",
        },
      ],
    },
  },
  "id": "73859F8D26679D2E52597326",
  "modelId": "testModel",
}
`;

exports[`services > file > uploadFile > success 1`] = `
{
  "_id": [MockFunction spy],
  "aggregate": [MockFunction spy],
  "avScan": {
    "aggregate": [MockFunction spy],
    "delete": [MockFunction spy],
    "find": [MockFunction spy] {
      "calls": [
        [
          {
            "fileId": [MockFunction spy],
          },
        ],
      ],
      "results": [
        {
          "type": "return",
          "value": [Circular],
        },
      ],
    },
    "findOne": [MockFunction spy],
  },
  "delete": [MockFunction spy],
  "find": [MockFunction spy],
  "findOneAndDelete": [MockFunction spy],
  "group": [MockFunction spy],
  "id": "function(...s) {
    let r = T(t);
    r.called = !0, r.callCount++, r.calls.push(s);
    let S = r.next.shift();
    if (S) {
      r.results.push(S);
      let [o, g] = S;
      if (o === "ok")
        return g;
      throw g;
    }
    let p, c = "ok", a = r.results.length;
    if (r.impl)
      try {
        new.target ? p = Reflect.construct(r.impl, s, new.target) : p = r.impl.apply(this, s), c = "ok";
      } catch (o) {
        throw p = o, c = "error", r.results.push([c, o]), o;
      }
    let R = [c, p];
    return w(p) && p.then(
      (o) => r.resolves[a] = ["ok", o],
      (o) => r.resolves[a] = ["error", o]
    ), r.results.push(R), p;
  }",
  "match": [MockFunction spy],
  "save": [MockFunction spy] {
    "calls": [
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": {
          "_id": [MockFunction spy],
          "aggregate": [MockFunction spy],
          "delete": [MockFunction spy],
          "find": [MockFunction spy],
          "findOneAndDelete": [MockFunction spy],
          "group": [MockFunction spy],
          "match": [MockFunction spy],
          "save": [MockFunction spy] {
            "calls": [
              [],
            ],
            "results": [
              {
                "type": "return",
                "value": [Circular],
              },
            ],
          },
          "size": 100,
          "toObject": [MockFunction spy] {
            "calls": [
              [],
            ],
            "results": [
              {
                "type": "return",
                "value": [Circular],
              },
            ],
          },
        },
      },
    ],
  },
  "size": 100,
  "toObject": [MockFunction spy] {
    "calls": [
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": {
          "_id": [MockFunction spy],
          "aggregate": [MockFunction spy],
          "delete": [MockFunction spy],
          "find": [MockFunction spy],
          "findOneAndDelete": [MockFunction spy],
          "group": [MockFunction spy],
          "match": [MockFunction spy],
          "save": [MockFunction spy] {
            "calls": [
              [],
            ],
            "results": [
              {
                "type": "return",
                "value": [Circular],
              },
            ],
          },
          "size": 100,
          "toObject": [MockFunction spy] {
            "calls": [
              [],
            ],
            "results": [
              {
                "type": "return",
                "value": [Circular],
              },
            ],
          },
        },
      },
    ],
  },
}
`;

exports[`services > file > uploadFile > virus scan initialised 1`] = `
{
  "_id": [MockFunction spy],
  "aggregate": [MockFunction spy],
  "avScan": {
    "aggregate": [MockFunction spy],
    "delete": [MockFunction spy],
    "find": [MockFunction spy] {
      "calls": [
        [
          {
            "fileId": [MockFunction spy],
          },
        ],
      ],
      "results": [
        {
          "type": "return",
          "value": [Circular],
        },
      ],
    },
    "findOne": [MockFunction spy],
  },
  "delete": [MockFunction spy],
  "find": [MockFunction spy],
  "findOneAndDelete": [MockFunction spy],
  "group": [MockFunction spy],
  "id": "function(...s) {
    let r = T(t);
    r.called = !0, r.callCount++, r.calls.push(s);
    let S = r.next.shift();
    if (S) {
      r.results.push(S);
      let [o, g] = S;
      if (o === "ok")
        return g;
      throw g;
    }
    let p, c = "ok", a = r.results.length;
    if (r.impl)
      try {
        new.target ? p = Reflect.construct(r.impl, s, new.target) : p = r.impl.apply(this, s), c = "ok";
      } catch (o) {
        throw p = o, c = "error", r.results.push([c, o]), o;
      }
    let R = [c, p];
    return w(p) && p.then(
      (o) => r.resolves[a] = ["ok", o],
      (o) => r.resolves[a] = ["error", o]
    ), r.results.push(R), p;
  }",
  "match": [MockFunction spy],
  "save": [MockFunction spy] {
    "calls": [
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": {
          "_id": [MockFunction spy],
          "aggregate": [MockFunction spy],
          "delete": [MockFunction spy],
          "find": [MockFunction spy],
          "findOneAndDelete": [MockFunction spy],
          "group": [MockFunction spy],
          "match": [MockFunction spy],
          "save": [MockFunction spy] {
            "calls": [
              [],
            ],
            "results": [
              {
                "type": "return",
                "value": [Circular],
              },
            ],
          },
          "size": 100,
          "toObject": [MockFunction spy] {
            "calls": [
              [],
            ],
            "results": [
              {
                "type": "return",
                "value": [Circular],
              },
            ],
          },
        },
      },
    ],
  },
  "size": 100,
  "toObject": [MockFunction spy] {
    "calls": [
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": {
          "_id": [MockFunction spy],
          "aggregate": [MockFunction spy],
          "delete": [MockFunction spy],
          "find": [MockFunction spy],
          "findOneAndDelete": [MockFunction spy],
          "group": [MockFunction spy],
          "match": [MockFunction spy],
          "save": [MockFunction spy] {
            "calls": [
              [],
            ],
            "results": [
              {
                "type": "return",
                "value": [Circular],
              },
            ],
          },
          "size": 100,
          "toObject": [MockFunction spy] {
            "calls": [
              [],
            ],
            "results": [
              {
                "type": "return",
                "value": [Circular],
              },
            ],
          },
        },
      },
    ],
  },
}
`;

exports[`services > file > uploadFile > virus scan initialised 2`] = `[]`;
